### PR TITLE
Add support for GASNet PSHM and send progress thread

### PR DESCRIPTION
**Pull request by jhh67**
This PR makes several changes to add support for GASNet PSHM (shared-memory bypass) for co-locales, and to improve RDMA performance via a send progress thread. These are related because the both require changes to how GASNet progress threads are implemented. The changes include:

- Modify the communication task functionality in `tasks-qthreads.c` to support more than one progress thread.
-  When there are co-locales, create an external GASNet progress thread to progress PSHM.
- When there is an external GASNet progress thread, do not create an internal RCV thread because it is redundant.
-  By default, set `GASNET_SND_THREAD` to create an internal GASNet thread that progresses send operations. This can be overridden by setting `GASNET_SND_THREAD=false` in the environment.
- Set `GASNET_SND_THREAD_POLL_MODE=exclusive`, which gives the GASNet send progress thread exclusive access to the NIC.
- If `CHPL_RT_COMM_GASNET_DEDICATED_PROGRESS_CORE=true`, reserve one core for internal and external GASNet progress threads and bind those threads to the core.
- If `CHPL_RT_COMM_GASNET_DEDICATED_PROGRESS_CORE=false`, bind the internal and external GASNet progress threads to the locale's accessible cores.
- In the `gasnet` shim, set `GASNET_SUPERNODE_MAXSIZE=1 `if there are no-colocales. This will prevent the use of PSHM, which cannot be used because the `gasnet` shim does not create an external GASNet progress thread.
- Interpose `comm_task_trampoline` between `chpl_task_createCommTask` and `chpl_task_wrapper`. `chpl_task_createCommTask` creates a thread that  invokes `comm_task_trampoline` which binds itself to the proper core(s) before creating another thread that invokes `chpl_task_wrapper`. This  ensures that the second thread is bound to the proper core(s) and its stack has the proper NUMA affinity. The thread created by `chpl_task_createCommTask` exits.

Closes https://github.com/Cray/chapel-private/issues/6369.
Closes https://github.com/Cray/chapel-private/issues/6368.
Closes https://github.com/Cray/chapel-private/issues/6228.

### CODE REVIEW

**Review by bonachea — CHANGES_REQUESTED** [review]
Hi @jhh67 - thanks for all this great work!! :-)

I don't claim to follow the full implications of the changes in chapel's task and topo modules, but I've added some initial constructive feedback based on source inspection of the comm/gasnet module changes where I'm more familiar. 

I realize this is still a draft PR, but we're very eager to see this work succeed so hopefully you find my suggestions helpful!

Regarding this statement in the commit message / PR description:
> By default, set `GASNET_SND_THREAD` to create an internal GASNet thread that progresses send operations. This can be overridden by setting `GASNET_SND_THREAD=false` in the environment.

Note the `false` setting suggested above doesn't actually work as written, and will generate the following error from GASNet:
```text
If used, environment variable 'GASNET_SND_THREAD' must be set to 'Y|YES|y|yes|1' or 'N|n|NO|no|0'
```

Better documentation would be `GASNET_SND_THREAD=0`.

File: runtime/src/tasks/qthreads/tasks-qthreads.c [hunk 1]
```diff
@@ -886,18 +953,14 @@ int chpl_task_createCommTask(chpl_fn_p fn,
...
-    return pthread_create(&chpl_qthread_comm_pthread,
-                          NULL, comm_task_wrapper, &wrapper_info);
+    comm_task_wrapper_info_t *wrapper_info;
+    wrapper_info = chpl_malloc(sizeof(*wrapper_info));
+    wrapper_info->fn = fn;
+    wrapper_info->arg = arg;
+    wrapper_info->cpu = cpu;
+    wrapper_info->thread = allocate_comm_task();
+    pthread_t dummy;
+    return pthread_create(&dummy, NULL, comm_task_trampoline, wrapper_info);
```
[the remark this answers is not in the document]
Comment (jhh67) [reply runtime/src/tasks/qthreads/tasks-qthreads.c:963]: Thanks for catching the error and suggesting a fix.

File: runtime/src/comm/gasnet/comm-gasnet.c [hunk 2]
```diff
@@ -868,6 +868,11 @@ void chpl_comm_init(int *argc_p, char ***argv_p) {
   setup_ibv();
   setup_polling();
 
+  // Setting this to 1 will disable PSHM even if it was enabled during
+  // configuration. PSHM requires an external progress thread, which this
+  // shim doesn't create.
+  chpl_env_set("GASNET_SUPERNODE_MAXSIZE", "1", 1);
```
[the remark this answers is not in the document]
Comment (jhh67) [reply runtime/src/comm/gasnet/comm-gasnet.c]: Fixed. I could not commit your suggestion directly because I rebased the underlying branch.

File: runtime/src/comm/gasnet/comm-gasnet-ex.c [hunk 3]
```diff
@@ -862,20 +884,27 @@ void chpl_comm_init(int *argc_p, char ***argv_p) {
...
+  // get information about co-locales
   gex_System_QueryHostInfo(NULL, &infoCount, &myIndex);
   chpl_set_num_locales_on_node((int32_t) infoCount);
   chpl_set_local_rank(myIndex);
 }
 
 void chpl_comm_pre_mem_init(void) {
+
+  if (chpl_env_rt_get_bool("COMM_GASNET_DEDICATED_PROGRESS_CORE", false)) {
+    reservedCore = chpl_topo_reserveCPUPhysical();
```
[the remark this answers is not in the document]
Comment (jhh67) [reply runtime/src/comm/gasnet/comm-gasnet-ex.c:906]: Yes. At the moment we do not reserve a core in the `ofi` comm layer, for now I'd like to do the same in `gasnet`. The impact is depends on the workload and whether or not hyper-threading is enabled. We can change the defaults down the road if we have enough benchmarking to indicate that `true` is a better option, and we'll have to give the users lots of warning.

File: runtime/src/comm/gasnet/comm-gasnet-ex.c [hunk 4]
```diff
@@ -751,11 +753,31 @@ static void polling(void* x) {
...
+  // start_gasnet_progress_threads we may decide not to create it.
+  chpl_env_set("GASNET_RCV_THREAD", "1", 1 /*overwrite*/);
+
+  // By default we want a GASNet send progress thread. Note that we do
+  // not override any existing value, so the user can disable it if desired.
+  chpl_env_set("GASNET_SND_THREAD", "1", 0 /*overwrite*/);
+
+  // If there is a send thread give it exclusive access to the NIC. This
+  // has no effect if there isn't a send thread.
+  chpl_env_set("GASNET_SND_THREAD_POLL_MODE", "exclusive", 1 /*overwrite*/);
```
[the remark this answers is not in the document]
Comment (jhh67) [reply runtime/src/comm/gasnet/comm-gasnet-ex.c:769]: I can go along this rationale, with the exception that we need to overwrite any user-provided setting that causes us to break. I don't think that's the case here, however.

File: runtime/src/comm/gasnet/comm-gasnet.c [hunk 5]
```diff
@@ -868,6 +868,19 @@ void chpl_comm_init(int *argc_p, char ***argv_p) {
...
+  if (!pollingRequired) {
+    // disable PSHM even if it was enabled during configuration.
+    chpl_env_set("GASNET_SUPERNODE_MAXSIZE", "1", 0);
+  }
+
+  // Setting this to 1 will disable PSHM even if it was enabled during
+  // configuration. PSHM requires an external progress thread, which this
+  // shim doesn't create.
+  chpl_env_set("GASNET_SUPERNODE_MAXSIZE", "1", 1);
+
```
Comment (bonachea) [inline runtime/src/comm/gasnet/comm-gasnet.c:883]: @jhh67 The merge conflict was resolved incorrectly and the result is two independent blocks doing contradictory things.

Please delete or revert the second block above.

```suggestion
```

**Review by bonachea — CHANGES_REQUESTED** [review]
Thanks for the updates @jhh67 !

New changes look good, with the exception of one git cherry-pick snafu that needs to be resolved.

Otherwise, I think this PR is ready to graduate out of "draft" status!

File: runtime/src/comm/gasnet/comm-gasnet-ex.c [hunk 6]
```diff
@@ -723,8 +725,10 @@ int32_t chpl_comm_getMaxThreads(void) {
 //
 static volatile int pollingRunning;
 static volatile int pollingQuit;
-static chpl_bool pollingRequired;
+static chpl_bool pollingRequired = false;
 static atomic_spinlock_t pollingLock;
+static chpl_bool usePSHM = false;
+static chpl_bool pshmInUse = false;
```
Comment (bonachea) [inline runtime/src/comm/gasnet/comm-gasnet-ex.c:731]: @jhh67 I strongly object to this change that you just folded into a rebase, which (given the current value of `usePSHM = false`) has the effect of disabling PSHM by default. I think this effectively defeats one of the primary goals of this PR, which is to provide fast co-locale comms by default.

Was this change unintentional?

```suggestion
static chpl_bool usePSHM = true;
static chpl_bool pshmInUse = false;
```

File: runtime/src/comm/gasnet/comm-gasnet-ex.c [hunk 7]
```diff
@@ -862,20 +892,36 @@ void chpl_comm_init(int *argc_p, char ***argv_p) {
...
   set_num_comm_domains();
   setup_ibv();
-  setup_polling();
 
-  GASNET_Safe(gex_Client_Init(&myclient, &myep, &myteam, "chapel", argc_p, argv_p, GEX_FLAG_USES_GASNET1));
+  usePSHM = chpl_env_rt_get_bool("COMM_GASNET_USE_PSHM", usePSHM);
+  if (!usePSHM) {
+    // Setting this to 1 will disable PSHM even if it was enabled during
+    // GASNet configuration.
+    chpl_env_set("GASNET_SUPERNODE_MAXSIZE", "1", 0);
```
Comment (bonachea) [inline runtime/src/comm/gasnet/comm-gasnet-ex.c:900]: On a related note, it seems problematic to introduce a `CHPL_COMM_GASNET_USE_PSHM` envvar that has overlapping behavior with `GASNET_SUPERNODE_MAXSIZE`. So for example if a user sets: 

1. Use case 1: `CHPL_COMM_GASNET_USE_PSHM=1 GASNET_SUPERNODE_MAXSIZE=1` or
2. Use case 2: `CHPL_COMM_GASNET_USE_PSHM=0 GASNET_SUPERNODE_MAXSIZE=0`, 

It's very non-obvious that these each represent conflicting inputs, and that the behavior will depend critically on the details of this conflict is handled by the backend.

**Review by bonachea — CHANGES_REQUESTED** [review]

File: runtime/src/comm/gasnet/comm-gasnet-ex.c [hunk 8]
```diff
@@ -946,8 +991,37 @@ int chpl_comm_run_in_lldb(int argc, char* argv[], int lldbArgnum, int* status) {
...
+                                 info[i].gex_progress_arg, reservedCore)) {
+      chpl_internal_error("unable to start internal GASNet progress thread");
+    }
+  }
+}
+
 void chpl_comm_post_task_init(void) {
   start_polling();
+  start_gasnet_progress_threads();
+#if defined(GASNET_CONDUIT_IBV)
```
Comment (PHHargrove) [inline runtime/src/comm/gasnet/comm-gasnet-ex.c:1015]: @jhh67 I see a couple issues with the `GASNET_SND_THREAD` portion of the verbose output below which I think diminish its value.

1. If the GASNet-EX configure arguments disable the send thread, then reporting a true value of the env var gives the "wrong answer".
2. If I change the default value to enable the send thread by default in a future release, then reporting for the unset case will also give the "wrong answer".

Instead of the `enabled` here, I suggest a `have_snd_thread` Boolean which starts `false`, but is set to `true` if the loop above observes `info[i].gex_thread_roles && GEX_THREAD_ROLE_SND`

File: runtime/src/comm/gasnet/comm-gasnet-ex.c [hunk 9]
```diff
@@ -946,8 +991,37 @@ int chpl_comm_run_in_lldb(int argc, char* argv[], int lldbArgnum, int* status) {
...
 
+static void start_gasnet_progress_threads(void) {
+  unsigned int count;
+  const gex_ProgressThreadInfo_t *info;
+  GASNET_Safe(gex_System_QueryProgressThreads(myclient, &count, &info, 0) );
+  for (int i = 0; i < count; i++) {
+
+    // Don't create an internal receive thread if there is an external one
+    if (pollingRequired &&
+        (info[i].gex_thread_roles == GEX_THREAD_ROLE_RCV)) {
```
Comment (PHHargrove) [inline runtime/src/comm/gasnet/comm-gasnet-ex.c:1002]: @jhh67 The line above tests for *equality* to a value which is documented as being just one bit among others which may be set.  Currently, the GASNet ibv-conduit implementation does not set both `GEX_THREAD_ROLE_RCV` and `GEX_THREAD_ROLE_SND`, but in the future it could.
Dan and I discussed this logic and believe the current test for equality is probably the correct one because *if* in the future there was a thread serving both the `RCV` and `SND` roles, you would want it to run for the `SND` activity.  However, we don't know if that was your intent or just the way things worked out.  So, an explanatory comment here would be useful.
TL;DR:  Request something like ` // Test for equality ensures a combination RCV+SND thread will still run`

File: runtime/src/comm/gasnet/comm-gasnet-ex.c (hunk 8 above)
Comment (bonachea) [reply runtime/src/comm/gasnet/comm-gasnet-ex.c:1015]: Note: Paul meant `&` above not `&&`.

Maybe also add an analogous `printf` for the recv thread?

File: runtime/src/comm/gasnet/comm-gasnet.c [hunk 10]
```diff
@@ -868,6 +868,14 @@ void chpl_comm_init(int *argc_p, char ***argv_p) {
   setup_ibv();
   setup_polling();
 
+  // PSHM needs an external progress thread to guarantee AM handlers make
+  // progress in the absence of other GASNet calls, so if we've disabled our
+  // external progress thread then we should disable PSHM
+  if (!pollingRequired) {
+    // disable PSHM even if it was enabled during configuration.
+    chpl_env_set("GASNET_SUPERNODE_MAXSIZE", "1", 0);
```
Comment (jabraham17) [inline runtime/src/comm/gasnet/comm-gasnet.c:876]: Should this also show a warning telling the user what happened so they are not suprised?

File: runtime/src/topo/hwloc/topo-hwloc.c [hunk 11]
```diff
@@ -785,7 +785,7 @@ int getCPUs(hwloc_cpuset_t cpuset, int *cpus, int size) {
 //
 int chpl_topo_getCPUs(chpl_bool physical, int *cpus, int count) {
   // Initializes CPU information.
-  okToReserveCPU = false;
+    cantReserveCPU = "chpl_topo_getCPUs called";
```
Comment (jabraham17) [inline runtime/src/topo/hwloc/topo-hwloc.c:788]: nit, extra spaces added

**Review by jabraham17 — APPROVED** [review]

File: runtime/src/comm/gasnet/comm-gasnet.c (hunk 10 above)
Comment (jhh67) [reply runtime/src/comm/gasnet/comm-gasnet.c:876]: I don't think so. It's not like the user requested PSHM and we are ignoring their request, it's that by default the `gasnet` library is now configured with PSHM enabled but we have to disable it at runtime in the `gasnet` shim because that shim doesn't create an external polling thread. Previously, PSHM was not enabled during configuration so this wasn't an issue.

**Review by bonachea — APPROVED** [review]
